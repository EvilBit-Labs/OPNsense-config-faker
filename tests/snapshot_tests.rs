//! Snapshot tests for OPNsense Config Faker
//!
//! These tests capture output snapshots to ensure CLI output remains consistent
//! across changes and updates.

mod common;

use common::{cli_command, create_temp_csv, create_temp_xml, normalize_output, TestOutputExt};
use insta::assert_snapshot;

/// Test CSV generation output with stable snapshots
#[test]
fn test_csv_output_snapshot() {
    let (temp_file, csv_path) = create_temp_csv(
        "snapshot_test_",
        &[
            &["VLAN", "IP Range", "Description"],
            &["100", "192.168.1.0/24", "Test Network"],
            &["200", "192.168.2.0/24", "Production"],
        ],
    )
    .unwrap();

    let output = cli_command()
        .arg("generate")
        .arg("--format")
        .arg("csv")
        .arg("--count")
        .arg("3")
        .arg("--output")
        .arg(&csv_path)
        .arg("--force") // Force overwrite existing temp file
        .arg("--seed")
        .arg("12345") // Fixed seed for reproducible output
        .run_success();

    // Normalize output for stable snapshots
    let normalized_stdout = normalize_output(&output.stdout);
    let normalized_stderr = normalize_output(&output.stderr);

    // Snapshot the normalized CLI output
    assert_snapshot!("csv_generation_success_stdout", normalized_stdout);
    assert_snapshot!("csv_generation_success_stderr", normalized_stderr);

    // Also snapshot the generated CSV content
    let csv_content = std::fs::read_to_string(&csv_path).unwrap();
    assert_snapshot!("generated_csv_content", csv_content);

    drop(temp_file);
}

/// Test help output snapshots
#[test]
fn test_help_output_snapshots() {
    // Test main help
    let output = cli_command().arg("--help").run_success();

    assert_snapshot!("main_help_output", output.normalized_stdout());

    // Test generate help
    let output = cli_command().arg("generate").arg("--help").run_success();

    assert_snapshot!("generate_help_output", output.normalized_stdout());

    // Test completions help
    let output = cli_command().arg("completions").arg("--help").run_success();

    assert_snapshot!("completions_help_output", output.normalized_stdout());
}

/// Test error output snapshots for consistent error messages
#[test]
fn test_error_output_snapshots() {
    // Test missing output file error
    let output = cli_command()
        .arg("generate")
        .arg("--format")
        .arg("csv")
        .arg("--count")
        .arg("5")
        .run_failure();

    assert_snapshot!("missing_csv_output_error", output.normalized_stderr());

    // Test missing XML base config error
    let output = cli_command()
        .arg("generate")
        .arg("--format")
        .arg("xml")
        .arg("--count")
        .arg("5")
        .run_failure();

    assert_snapshot!("missing_xml_base_config_error", output.normalized_stderr());

    // Test file exists without force error
    let (temp_file, csv_path) = create_temp_csv("error_test_", &[&["existing", "data"]]).unwrap();

    let output = cli_command()
        .arg("generate")
        .arg("--format")
        .arg("csv")
        .arg("--count")
        .arg("2")
        .arg("--output")
        .arg(&csv_path)
        .arg("--seed")
        .arg("42")
        .run_failure();

    assert_snapshot!(
        "file_exists_without_force_error",
        output.normalized_stderr()
    );

    drop(temp_file);
}

/// Test deprecated command warnings
#[test]
fn test_deprecated_command_snapshots() {
    let (temp_file, csv_path) =
        create_temp_csv("deprecated_test_", &[&["header1", "header2"]]).unwrap();

    // Test deprecated CSV command
    let output = cli_command()
        .arg("csv")
        .arg("--count")
        .arg("3")
        .arg("--output")
        .arg(&csv_path)
        .arg("--seed")
        .arg("42")
        .run_failure();

    // Capture both stdout and stderr for deprecation messages
    assert_snapshot!("deprecated_csv_stdout", output.normalized_stdout());
    assert_snapshot!("deprecated_csv_stderr", output.normalized_stderr());

    drop(temp_file);
}

/// Test XML generation output snapshots
#[test]
fn test_xml_generation_snapshots() {
    let base_xml_content = r#"<?xml version="1.0"?>
<opnsense version="24.1" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
  <system>
    <hostname>opnsense</hostname>
  </system>
  <interfaces>
    <lan>
      <if>em0</if>
      <ipaddr>dhcp</ipaddr>
    </lan>
    <wan>
      <if>em1</if>
      <ipaddr>dhcp</ipaddr>
    </wan>
  </interfaces>
</opnsense>"#;

    let (temp_xml_file, xml_path) = create_temp_xml("base_config_", base_xml_content).unwrap();
    let (temp_csv_file, csv_path) = create_temp_csv(
        "vlan_data_",
        &[
            &["VLAN", "IP Range", "Beschreibung", "WAN"],
            &["100", "192.168.100.0/24", "Development", "1"],
            &["200", "192.168.200.0/24", "Testing", "2"],
        ],
    )
    .unwrap();

    // Test XML generation using CSV input
    let output = cli_command()
        .arg("generate")
        .arg("--format")
        .arg("xml")
        .arg("--base-config")
        .arg(&xml_path)
        .arg("--csv-file")
        .arg(&csv_path)
        .arg("--output-dir")
        .arg("/tmp") // Ensure consistent output directory for snapshots
        .arg("--force") // Force overwrite existing files
        .run_success();

    assert_snapshot!("xml_generation_with_csv_stdout", output.normalized_stdout());
    assert_snapshot!("xml_generation_with_csv_stderr", output.normalized_stderr());

    drop(temp_xml_file);
    drop(temp_csv_file);
}

/// Test progress output normalization
#[test]
fn test_progress_normalization() {
    let (temp_file, csv_path) = create_temp_csv("progress_test_", &[]).unwrap();

    // Remove temp file first to avoid force requirement
    drop(temp_file);
    std::fs::remove_file(&csv_path).ok(); // Ignore error if file doesn't exist

    let output = cli_command()
        .arg("generate")
        .arg("--format")
        .arg("csv")
        .arg("--count")
        .arg("10") // Larger count to trigger progress indicators
        .arg("--output")
        .arg(&csv_path)
        .arg("--seed")
        .arg("999")
        .run_success();

    // The normalize_output function should remove progress indicators
    let normalized = normalize_output(&output.stdout);

    // Verify progress characters are removed
    assert!(!normalized.contains("⠁"));
    assert!(!normalized.contains("⠂"));
    assert!(!normalized.contains("⠄"));
    assert!(!normalized.contains("▪"));
    assert!(!normalized.contains("▫"));

    assert_snapshot!("progress_normalized_output", normalized);
}

/// Test bash completion script snapshot
#[test]
fn test_bash_completion_snapshot() {
    let output = cli_command().arg("completions").arg("bash").run_success();

    // Bash completions should be consistent
    let normalized = normalize_output(&output.stdout);
    assert_snapshot!("bash_completion_script", normalized);
}

/// Test zsh completion script snapshot
#[test]
fn test_zsh_completion_snapshot() {
    let output = cli_command().arg("completions").arg("zsh").run_success();

    let normalized = normalize_output(&output.stdout);
    assert_snapshot!("zsh_completion_script", normalized);
}

/// Test fish completion script snapshot
#[test]
fn test_fish_completion_snapshot() {
    let output = cli_command().arg("completions").arg("fish").run_success();

    let normalized = normalize_output(&output.stdout);
    assert_snapshot!("fish_completion_script", normalized);
}
