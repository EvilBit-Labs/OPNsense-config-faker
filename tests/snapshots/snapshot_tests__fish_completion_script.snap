---
source: tests/snapshot_tests.rs
expression: normalized
---
# Print an optspec for argparse to handle cmd's options that are independent of any subcommand. function __fish_opnsense_config_faker_global_optspecs string join \n h/help V/version end function __fish_opnsense_config_faker_needs_command # Figure out if the current invocation already has a command. set -l cmd (commandline -opc) set -e cmd[1] argparse -s (__fish_opnsense_config_faker_global_optspecs) -- $cmd 2>/dev/null or return if set -q argv[1] # Also print the command, so this can be used to figure out what it is. echo $argv[1] return 1 end return 0 end function __fish_opnsense_config_faker_using_subcommand set -l cmd (__fish_opnsense_config_faker_needs_command) test -z "$cmd" and return 1 contains -- $cmd[1] $argv end complete -c opnsense-config-faker -n "__fish_opnsense_config_faker_needs_command" -s h -l help -d 'Print help' complete -c opnsense-config-faker -n "__fish_opnsense_config_faker_needs_command" -s V -l version -d 'Print version' complete -c opnsense-config-faker -n "__fish_opnsense_config_faker_needs_command" -f -a "generate" -d 'Generate network configuration data in CSV or XML format' complete -c opnsense-config-faker -n "__fish_opnsense_config_faker_needs_command" -f -a "completions" -d 'Generate shell completions for the specified shell' complete -c opnsense-config-faker -n "__fish_opnsense_config_faker_needs_command" -f -a "csv" -d 'DEPRECATED: Use \'generate --format csv\' instead' complete -c opnsense-config-faker -n "__fish_opnsense_config_faker_needs_command" -f -a "xml" -d 'DEPRECATED: Use \'generate --format xml\' instead' complete -c opnsense-config-faker -n "__fish_opnsense_config_faker_needs_command" -f -a "help" -d 'Print this message or the help of the given subcommand(s)' complete -c opnsense-config-faker -n "__fish_opnsense_config_faker_using_subcommand generate" -s f -l format -d 'Output format (csv or xml)' -r -f -a "csv\t'Generate CSV file with VLAN configuration data' xml\t'Generate complete OPNsense XML configuration'" complete -c opnsense-config-faker -n "__fish_opnsense_config_faker_using_subcommand generate" -s c -l count -d 'Number of VLAN configurations to generate' -r complete -c opnsense-config-faker -n "__fish_opnsense_config_faker_using_subcommand generate" -s o -l output -d 'Output file path (for CSV format) or directory (for XML format)' -r -F complete -c opnsense-config-faker -n "__fish_opnsense_config_faker_using_subcommand generate" -l output-dir -d 'Output directory for generated XML files (XML format only)' -r -F complete -c opnsense-config-faker -n "__fish_opnsense_config_faker_using_subcommand generate" -s b -l base-config -d 'Base OPNsense configuration XML file (required for XML format)' -r -F complete -c opnsense-config-faker -n "__fish_opnsense_config_faker_using_subcommand generate" -l csv-file -d 'Use existing CSV file for configuration data (XML format only)' -r -F complete -c opnsense-config-faker -n "__fish_opnsense_config_faker_using_subcommand generate" -l firewall-nr -d 'Firewall number for naming (used in filenames for XML format)' -r complete -c opnsense-config-faker -n "__fish_opnsense_config_faker_using_subcommand generate" -l opt-counter -d 'OPT interface counter starting value (XML format only)' -r complete -c opnsense-config-faker -n "__fish_opnsense_config_faker_using_subcommand generate" -l seed -d 'Random seed for reproducible generation' -r complete -c opnsense-config-faker -n "__fish_opnsense_config_faker_using_subcommand generate" -s F -l force -d 'Force overwrite existing files' complete -c opnsense-config-faker -n "__fish_opnsense_config_faker_using_subcommand generate" -l no-color -d 'Disable colored output (useful for scripts and CI)' complete -c opnsense-config-faker -n "__fish_opnsense_config_faker_using_subcommand generate" -s i -l interactive -d 'Interactive mode - prompt for missing required arguments' complete -c opnsense-config-faker -n "__fish_opnsense_config_faker_using_subcommand generate" -s h -l help -d 'Print help (see more with \'--help\')' complete -c opnsense-config-faker -n "__fish_opnsense_config_faker_using_subcommand completions" -s h -l help -d 'Print help' complete -c opnsense-config-faker -n "__fish_opnsense_config_faker_using_subcommand csv" -s c -l count -d 'Number of VLAN configurations to generate' -r complete -c opnsense-config-faker -n "__fish_opnsense_config_faker_using_subcommand csv" -s o -l output -d 'Output CSV file path' -r -F complete -c opnsense-config-faker -n "__fish_opnsense_config_faker_using_subcommand csv" -l seed -d 'Random seed for reproducible generation' -r complete -c opnsense-config-faker -n "__fish_opnsense_config_faker_using_subcommand csv" -s f -l force -d 'Force overwrite existing files' complete -c opnsense-config-faker -n "__fish_opnsense_config_faker_using_subcommand csv" -s h -l help -d 'Print help (see more with \'--help\')' complete -c opnsense-config-faker -n "__fish_opnsense_config_faker_using_subcommand xml" -s b -l base-config -d 'Base OPNsense configuration XML file' -r -F complete -c opnsense-config-faker -n "__fish_opnsense_config_faker_using_subcommand xml" -s c -l count -d 'Number of VLAN configurations to generate (if not using CSV)' -r complete -c opnsense-config-faker -n "__fish_opnsense_config_faker_using_subcommand xml" -l csv-file -d 'Use existing CSV file for configuration data' -r -F complete -c opnsense-config-faker -n "__fish_opnsense_config_faker_using_subcommand xml" -s o -l output-dir -d 'Output directory for generated XML files' -r -F complete -c opnsense-config-faker -n "__fish_opnsense_config_faker_using_subcommand xml" -l firewall-nr -d 'Firewall number for naming (used in filenames)' -r complete -c opnsense-config-faker -n "__fish_opnsense_config_faker_using_subcommand xml" -l opt-counter -d 'OPT interface counter starting value' -r complete -c opnsense-config-faker -n "__fish_opnsense_config_faker_using_subcommand xml" -l seed -d 'Random seed for reproducible generation' -r complete -c opnsense-config-faker -n "__fish_opnsense_config_faker_using_subcommand xml" -s f -l force -d 'Force overwrite existing files' complete -c opnsense-config-faker -n "__fish_opnsense_config_faker_using_subcommand xml" -s h -l help -d 'Print help (see more with \'--help\')' complete -c opnsense-config-faker -n "__fish_opnsense_config_faker_using_subcommand help; and not __fish_seen_subcommand_from generate completions csv xml help" -f -a "generate" -d 'Generate network configuration data in CSV or XML format' complete -c opnsense-config-faker -n "__fish_opnsense_config_faker_using_subcommand help; and not __fish_seen_subcommand_from generate completions csv xml help" -f -a "completions" -d 'Generate shell completions for the specified shell' complete -c opnsense-config-faker -n "__fish_opnsense_config_faker_using_subcommand help; and not __fish_seen_subcommand_from generate completions csv xml help" -f -a "csv" -d 'DEPRECATED: Use \'generate --format csv\' instead' complete -c opnsense-config-faker -n "__fish_opnsense_config_faker_using_subcommand help; and not __fish_seen_subcommand_from generate completions csv xml help" -f -a "xml" -d 'DEPRECATED: Use \'generate --format xml\' instead' complete -c opnsense-config-faker -n "__fish_opnsense_config_faker_using_subcommand help; and not __fish_seen_subcommand_from generate completions csv xml help" -f -a "help" -d 'Print this message or the help of the given subcommand(s)'
